schemachanger: support named range zone config

annrpom opened the pull request:
### schemachanger: add RangeZoneConfig element
This patch adds a RangeConfig element to help support configuring
the zone of a named zone. This includes adding a `WalkNamedRanges` function
for decomposition.

Release note: None

---

### scbuild: handle pseudo-table ids in ensureDescriptor
For pseudo-table ids, we don't have a "real" descriptor.
This patch expands `ensureDescriptor` to account for this
by creating a cachedDesc that has everything but an actual
descriptor.

Epic: none

Release note: None

---

### schemachanger: add opgen rule for named range zc
This patch adds an opgen rule to support our NamedRangeZoneConfig
element.

Release note: None

---

### schemachanger: support named range zone config
This patch adds the functionality to configure
a zone configuration on a named range.

Fixes: https://github.com/cockroachdb/cockroach/issues/129891

Release note: None

annrpom committed:
schemachanger: add RangeZoneConfig element

This patch adds a RangeConfig element to help support configuring
the zone of a named zone. This includes adding a `WalkNamedRanges` function
for decomposition.

Release note: None

annrpom committed:
schemachanger: add opgen rule for named range zc

This patch adds an opgen rule to support our NamedRangeZoneConfig
element.

Release note: None

annrpom committed:
schemachanger: support named range zone config

This patch adds the functionality to configure
a zone configuration on a named range.

Fixes: #129891

Release note: None

annrpom committed:
scbuild: handle pseudo-table ids in ensureDescriptor

For pseudo-table ids, we don't have a "real" descriptor.
This patch expands `ensureDescriptor` to account for this
by creating a cachedDesc that has everything but an actual
descriptor.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/133094)
<!-- Reviewable:end -->

annrpom commented on pkg/sql/schemachanger/testdata/end_to_end/alter_named_range_configure_zone/alter_named_range_configure_zone.explain:
```diff
@@ -0,0 +1,23 @@
+ │         ├── 1 element transitioning toward PUBLIC
+ │         │    └── ABSENT → PUBLIC NamedRangeZoneConfig:{DescID: 0 (#0), SeqNum: 1}
+ │         └── 1 Mutation operation
+ │              └── AddNamedRangeZoneConfig {"ZoneConfig":{"nullVoterConstraintsIsEmpty":true,"numReplicas":7,"rangeMaxBytes":"536870912","rangeMinBytes":"134217728"}}
```
this looks weird and i'm not too sure of where this comes from...yet

spilchen reviewed (changes requested):
I did a pass through. It seems like the correct approach. I just had a few questions/comments.

> Reviewed 2 of 2 files at r1, 8 of 8 files at r2, 2 of 2 files at r3, 6 of 6 files at r4, 24 of 24 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-:-O9zZU7V87rnFcMfHmHx:b-mfh4up)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom)

___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 53 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-O9zPBFs3LzRdwjRfYFh:-O9zPBFt36A5Ni-mOpt8:b-cq0fpf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbcf3a6c09e14f7cb9872210c1c2d828171b0b29/pkg/sql/schemachanger/scbuild/builder_state.go#L53)):*
> ```Go
> // QueryByID implements the scbuildstmt.BuilderState interface.
> func (b *builderState) QueryByID(id catid.DescID) scbuildstmt.ElementResultSet {
> 	// TODO a bit weird here since it overlaps with invalid desc id
> ```

Why was this code block commented out? Is it related to pseudo-table IDs used in named ranges? Should we have the lookup for `zonepb.NamedZonesByID` first?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go` line 1062 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-O9zYncbEfVXUtgmreHO:-O9zYncbEfVXUtgmreHP:b-oq8k3v) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbcf3a6c09e14f7cb9872210c1c2d828171b0b29/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go#L1062)):*
> ```Go
> 		// memory default value.
> 		if keys.RootNamespaceID == uint32(obj.getTargetID()) {
> 			// TODO(reviewer callout): is it fine to use this? or do we want to
> ```

Pulling it from the planner seems like the right approach. There might be code, possibly in tests, that sets up the default zone config in a specific way.
___
*[`pkg/sql/schemachanger/scop/immediate_mutation.go` line 995 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-O9zVwHpBDZkMs-ZaBJx:-O9zVwHq41cawWaPsgtw:bywzy4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbcf3a6c09e14f7cb9872210c1c2d828171b0b29/pkg/sql/schemachanger/scop/immediate_mutation.go#L995)):*
> ```Go
> 	immediateMutationOp
> 	RangeID    descpb.ID
> 	ZoneConfig *zonepb.ZoneConfig
> ```

Should this be a pointer? It might be safer to perform a deep copy of the struct, which would help prevent accidental changes to the zone config.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 121 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-O9zU-K56_xzcRx94i8E:-O9zU-K56_xzcRx94i8F:bwzexoy) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbcf3a6c09e14f7cb9872210c1c2d828171b0b29/pkg/sql/schemachanger/scpb/elements.proto#L121)):*
> ```proto
>     Owner owner = 61 [(gogoproto.moretags) = "parent:\"Table, View, Sequence, Database, Schema, AliasType, EnumType\""];
>     UserPrivileges user_privileges = 62 [(gogoproto.moretags) = "parent:\"Table, View, Sequence, Database, Schema, AliasType, EnumType\""];
>     // TODO should i add a top-level element for this?
> ```

It seems like its a standalone object, so I think it need's a top-level element.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 122 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-O9zTJXQ29HHqqjIUXL0:-O9zTJXQ29HHqqjIUXL1:becejp6) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbcf3a6c09e14f7cb9872210c1c2d828171b0b29/pkg/sql/schemachanger/scpb/elements.proto#L122)):*
> ```proto
>     UserPrivileges user_privileges = 62 [(gogoproto.moretags) = "parent:\"Table, View, Sequence, Database, Schema, AliasType, EnumType\""];
>     // TODO should i add a top-level element for this?
>     NamedRangeZoneConfig named_range_zone_config = 63 [(gogoproto.moretags) = "parent:\"Table, View, Sequence, Database, Schema, AliasType, EnumType\""];
> ```

Commenting on the parent and placement with the common elements, do we have a named range zone config for each table/view/sequence/etc? There isn't an ID in the range that ties them together. So, I'm a little confused about this part.
___
*[`pkg/sql/schemachanger/screl/scalars.go` line 126 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-O9zNJGz0AMBry3nDHU6:-O9zNJGz0AMBry3nDHU7:bvzl8p1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9a941206e1e9a1e89ae7a9d3171836674dcdce9/pkg/sql/schemachanger/screl/scalars.go#L126)):*
> ```Go
> 		return version.IsActive(clusterversion.V24_2)
> 	case *scpb.ColumnComputeExpression, *scpb.FunctionSecurity, *scpb.LDRJobIDs,
> 		*scpb.NamedRangeZoneConfig, *scpb.PartitionZoneConfig, *scpb.Trigger, *scpb.TriggerName,
> ```

Should this be added in a separate case statement? We'll want to use `clusterversion.V25_1` once it is available.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/testdata/deprules` line 0 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-O9zNWOi0rQcpNQJEIQm:-O9zNWOi0rQcpNQJEIQn:biym820) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9a941206e1e9a1e89ae7a9d3171836674dcdce9/pkg/sql/schemachanger/scplan/internal/rules/current/testdata/deprules#L0)):*
When do we make a copy of the deprules? We need to preserve the 24.3 rules at some point. Can we still do this at some point in the future?
___
*[`pkg/sql/schemachanger/testdata/end_to_end/alter_named_range_configure_zone/alter_named_range_configure_zone.explain` line 12 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/133094#gh-1811391498:-O9zVJTl2vRru73m9Mia:bm65nis) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ce75c34e8e26aaba0c79197a9609f9fb5beae07/pkg/sql/schemachanger/testdata/end_to_end/alter_named_range_configure_zone/alter_named_range_configure_zone.explain#L12)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

this looks weird and i'm not too sure of where this comes from...yet

> this looks weird

by "this" i mean specifically why AddNamedRangeZoneConfig is printing that
</blockquote></details>

Agreed. It looks like it is printing out whatever the contents of the `AddNamedRangeZoneConfig` struct. 


<!-- Sent from Reviewable.io -->

annrpom committed:
schemachanger: add DefaultZoneConfig to builder

This patch plumbs `execCfg`'s `DefaultZoneConfig`
down to the schemachanger builder.

Epic: none

Release note: None

annrpom commented:
i found where it comes from https://github.com/cockroachdb/cockroach/blob/9f61a38b670e57873509568f67bb665e50cd1935/pkg/sql/schemachanger/scplan/plan_explain.go#L299-L305

yeah; it's printing out the zone config in there instead of the rangeID (like the AddDatabaseZoneConfig op) does since in this case, the rangeID is 0

i think one thing i can do is make a custom wrapper for the uint32 type in the elements proto file; this way, the 0 value cannnot be omitted

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133094)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 53 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-O9zPBFs3LzRdwjRfYFh:-OA-934LDNemQxQlFc9T:b-al2pke) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbcf3a6c09e14f7cb9872210c1c2d828171b0b29/pkg/sql/schemachanger/scbuild/builder_state.go#L53)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Why was this code block commented out? Is it related to pseudo-table IDs used in named ranges? Should we have the lookup for `zonepb.NamedZonesByID` first?
</blockquote></details>

let me delete it -- i wanted to see if CI was red with this change as i wasn't sure whether or not we expect to have some call on `QueryByID(0)` for one reason or the other. it looks like we don't. 

the problem is that the id associated with the default range is 0; this also happens to be the same value as `catid.InvalidDescID`. so now the workaround is that 0 gets associated with our named zone in `ensureDescriptor`
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go` line 1062 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-O9zYncbEfVXUtgmreHO:-OA-BysL6ZelS_obWp0c:bixyowy) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbcf3a6c09e14f7cb9872210c1c2d828171b0b29/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go#L1062)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Pulling it from the planner seems like the right approach. There might be code, possibly in tests, that sets up the default zone config in a specific way.
</blockquote></details>

ah tests, right; done
___
*[`pkg/sql/schemachanger/scop/immediate_mutation.go` line 995 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-O9zVwHpBDZkMs-ZaBJx:-OANjrXi27y4jzB1bpN0:bp6ri7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbcf3a6c09e14f7cb9872210c1c2d828171b0b29/pkg/sql/schemachanger/scop/immediate_mutation.go#L995)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Should this be a pointer? It might be safer to perform a deep copy of the struct, which would help prevent accidental changes to the zone config.
</blockquote></details>

is the point so that we can isolate changes made from immediateState to what we have in the builder? if so, this makes sense; we now`protutil.Clone` the zone config from our builder elements. let me know we are not on the same page!
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 121 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-O9zU-K56_xzcRx94i8E:-OA-JbCdAsqndWxhInoZ:brrwdwz) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbcf3a6c09e14f7cb9872210c1c2d828171b0b29/pkg/sql/schemachanger/scpb/elements.proto#L121)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

It seems like its a standalone object, so I think it need's a top-level element.
</blockquote></details>

the only weird thing is that top-level elements are defined as the following:
```
    // Top-level elements.
    // A.k.a descriptor-elements.
    // These elements own a corresponding descriptor in the catalog.
    Database database = 1;
    Schema schema = 2;
    View view = 3;
    Sequence sequence = 4;
    Table table = 5;
    EnumType enum_type = 6;
    AliasType alias_type = 7;
    CompositeType composite_type = 8;
    Function function = 9;
```
i'm not sure how much i am violating the DSC by putting a descriptorless element there. i'll just put this up near top-level elements, but make it more like a "zero-level"(? cred to rafi on name) element
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 122 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-O9zTJXQ29HHqqjIUXL0:-OA-JAK78q0FRG5VWYyn:b-k1qwdq) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbcf3a6c09e14f7cb9872210c1c2d828171b0b29/pkg/sql/schemachanger/scpb/elements.proto#L122)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Commenting on the parent and placement with the common elements, do we have a named range zone config for each table/view/sequence/etc? There isn't an ID in the range that ties them together. So, I'm a little confused about this part.
</blockquote></details>

we do not -- i actually just wasn't sure what to do about this part since i was not sure how this fit into top-level elems yet, but see above
___
*[`pkg/sql/schemachanger/screl/scalars.go` line 126 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-O9zNJGz0AMBry3nDHU6:-OA-KerZ7EZcc5zTv7kd:bhw4hc6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9a941206e1e9a1e89ae7a9d3171836674dcdce9/pkg/sql/schemachanger/screl/scalars.go#L126)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Should this be added in a separate case statement? We'll want to use `clusterversion.V25_1` once it is available.
</blockquote></details>

right -- thanks for the catch! i thought i was able to get it into 24.3 T_T. i'll probably let this PR sit until 25_1 is in
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/testdata/deprules` line  at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-O9zNWOi0rQcpNQJEIQm:-OA-Tu7o7yQaIUq3Du6R:bcumrvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9a941206e1e9a1e89ae7a9d3171836674dcdce9/pkg/sql/schemachanger/scplan/internal/rules/current/testdata/deprules#L)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

When do we make a copy of the deprules? We need to preserve the 24.3 rules at some point. Can we still do this at some point in the future?
</blockquote></details>

done/resolved -- for documentation: we do not need to preserve the 24_3 deprules here, but we can for the sake of getting it out of the way (so we did)
___
*[`pkg/sql/schemachanger/testdata/end_to_end/alter_named_range_configure_zone/alter_named_range_configure_zone.explain` line 12 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/133094#gh-1811391498:-OANyuE76_eAFeFluGdi:b9o02a9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ce75c34e8e26aaba0c79197a9609f9fb5beae07/pkg/sql/schemachanger/testdata/end_to_end/alter_named_range_configure_zone/alter_named_range_configure_zone.explain#L12)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Agreed. It looks like it is printing out whatever the contents of the `AddNamedRangeZoneConfig` struct.
</blockquote></details>

There are two ways I can go about this; i can fix this for the default range if i do the following:
```
--- a/pkg/sql/schemachanger/scplan/scviz/BUILD.bazel
+++ b/pkg/sql/schemachanger/scplan/scviz/BUILD.bazel
@@ -6,6 +6,7 @@ go_library(
     importpath = "github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scplan/scviz",
     visibility = ["//visibility:public"],
     deps = [
+        "//pkg/sql/schemachanger/scop",
         "//pkg/util/protoutil",
         "@com_github_cockroachdb_errors//:errors",
         "@com_github_gogo_protobuf//jsonpb",
diff --git a/pkg/sql/schemachanger/scplan/scviz/maps.go b/pkg/sql/schemachanger/scplan/scviz/maps.go
index 9f885582a35..ad10af50c19 100644
--- a/pkg/sql/schemachanger/scplan/scviz/maps.go
+++ b/pkg/sql/schemachanger/scplan/scviz/maps.go
@@ -12,6 +12,7 @@ import (
        "reflect"
        "regexp"
 
+       "github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scop"
        "github.com/cockroachdb/cockroach/pkg/util/protoutil"
        "github.com/cockroachdb/errors"
        "github.com/gogo/protobuf/jsonpb"
@@ -55,9 +56,14 @@ func ToMap(v interface{}, emitDefaults bool) (interface{}, error) {
        m := make(map[string]interface{}, vt.NumField())
        for i := 0; i < vt.NumField(); i++ {
                vvf := vv.Field(i)
-               if !vvf.CanInterface() || vvf.IsZero() {
+               if !vvf.CanInterface() {
                        continue
                }
+               if vvf.IsZero() {
+                       if _, ok := v.(*scop.AddNamedRangeZoneConfig); !ok {
+                               continue
+                       }
+               }
                var err error
                if m[vt.Field(i).Name], err = ToMap(vvf.Interface(), emitDefaults); err != nil {
                        return nil, err
```

or we can also reference the name:

```
AddNamedRangeZoneConfig {"RangeName":"default"}
```

i actually think i like the latter better, but i'm welcome to discussing


<!-- Sent from Reviewable.io -->

spilchen reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for the revisions.

> Reviewed 7 of 8 files at r9, 42 of 42 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-:-OAU2e379D4w_K8Bw4-7:b-ffsmf0)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @annrpom and @fqazi)

___
*[`pkg/sql/schemachanger/scdeps/sctestdeps/test_deps.go` line 1512 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-OATh6K9E4nT-v4nBuPt:-OATh6KA9ek8uGJ5wxKW:b-ihmi7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/f489804b5ef373bfc405b6049f1f3e1b8fa52bd0/pkg/sql/schemachanger/scdeps/sctestdeps/test_deps.go#L1512)):*
> ```Go
> 
> func (s *TestState) GetDefaultZoneConfig() *zonepb.ZoneConfig {
> 	return nil
> ```

Does `nil` work with `protoutil.Clone` (that's the only caller of this function)? Or does it matter.
___
*[`pkg/sql/schemachanger/scop/immediate_mutation.go` line 995 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-O9zVwHpBDZkMs-ZaBJx:-OAU2PcpCFdmqd_ACd3F:bnilc1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbcf3a6c09e14f7cb9872210c1c2d828171b0b29/pkg/sql/schemachanger/scop/immediate_mutation.go#L995)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

is the point so that we can isolate changes made from immediateState to what we have in the builder? if so, this makes sense; we now`protutil.Clone` the zone config from our builder elements. let me know we are not on the same page!
</blockquote></details>

Yes, this was my concern. Thanks for changing it.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 121 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-O9zU-K56_xzcRx94i8E:-OATi4f_8LOMvI9wBjMn:b-x7yc1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbcf3a6c09e14f7cb9872210c1c2d828171b0b29/pkg/sql/schemachanger/scpb/elements.proto#L121)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

the only weird thing is that top-level elements are defined as the following:
```
    // Top-level elements.
    // A.k.a descriptor-elements.
    // These elements own a corresponding descriptor in the catalog.
    Database database = 1;
    Schema schema = 2;
    View view = 3;
    Sequence sequence = 4;
    Table table = 5;
    EnumType enum_type = 6;
    AliasType alias_type = 7;
    CompositeType composite_type = 8;
    Function function = 9;
```
i'm not sure how much i am violating the DSC by putting a descriptorless element there. i'll just put this up near top-level elements, but make it more like a "zero-level"(? cred to rafi on name) element
</blockquote></details>

I'm good with name "zero-level elements".
___
*[`pkg/sql/schemachanger/screl/scalars.go` line 126 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-O9zNJGz0AMBry3nDHU6:-OAU-ogaEm302OWoLZvA:b-wu2kr9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9a941206e1e9a1e89ae7a9d3171836674dcdce9/pkg/sql/schemachanger/screl/scalars.go#L126)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

right -- thanks for the catch! i thought i was able to get it into 24.3 T_T. i'll probably let this PR sit until 25_1 is in
</blockquote></details>

I'm not sure when the `25_1` variable will be added, so I'm fine with the `TODO` you have for now if you want to get this in sooner.
___
*[`pkg/sql/schemachanger/testdata/end_to_end/alter_named_range_configure_zone/alter_named_range_configure_zone.explain` line 12 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/133094#gh-1811391498:-OATbYFVEPL_qA79KE8P:bnzwzlx) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ce75c34e8e26aaba0c79197a9609f9fb5beae07/pkg/sql/schemachanger/testdata/end_to_end/alter_named_range_configure_zone/alter_named_range_configure_zone.explain#L12)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

There are two ways I can go about this; i can fix this for the default range if i do the following:
```
--- a/pkg/sql/schemachanger/scplan/scviz/BUILD.bazel
+++ b/pkg/sql/schemachanger/scplan/scviz/BUILD.bazel
@@ -6,6 +6,7 @@ go_library(
     importpath = "github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scplan/scviz",
     visibility = ["//visibility:public"],
     deps = [
+        "//pkg/sql/schemachanger/scop",
         "//pkg/util/protoutil",
         "@com_github_cockroachdb_errors//:errors",
         "@com_github_gogo_protobuf//jsonpb",
diff --git a/pkg/sql/schemachanger/scplan/scviz/maps.go b/pkg/sql/schemachanger/scplan/scviz/maps.go
index 9f885582a35..ad10af50c19 100644
--- a/pkg/sql/schemachanger/scplan/scviz/maps.go
+++ b/pkg/sql/schemachanger/scplan/scviz/maps.go
@@ -12,6 +12,7 @@ import (
        "reflect"
        "regexp"
 
+       "github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scop"
        "github.com/cockroachdb/cockroach/pkg/util/protoutil"
        "github.com/cockroachdb/errors"
        "github.com/gogo/protobuf/jsonpb"
@@ -55,9 +56,14 @@ func ToMap(v interface{}, emitDefaults bool) (interface{}, error) {
        m := make(map[string]interface{}, vt.NumField())
        for i := 0; i < vt.NumField(); i++ {
                vvf := vv.Field(i)
-               if !vvf.CanInterface() || vvf.IsZero() {
+               if !vvf.CanInterface() {
                        continue
                }
+               if vvf.IsZero() {
+                       if _, ok := v.(*scop.AddNamedRangeZoneConfig); !ok {
+                               continue
+                       }
+               }
                var err error
                if m[vt.Field(i).Name], err = ToMap(vvf.Interface(), emitDefaults); err != nil {
                        return nil, err
```

or we can also reference the name:

```
AddNamedRangeZoneConfig {"RangeName":"default"}
```

i actually think i like the latter better, but i'm welcome to discussing
</blockquote></details>

I like the latter too. Let's leave `scviz` as is. Thanks for tracking this down!


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  Just one minor question

> Reviewed 2 of 2 files at r1, 7 of 8 files at r9, 42 of 42 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-:-OAXxhfnE1jA4b35rkTY:b-yxwu2a)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @annrpom)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 103 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/133094#-OAXwpunAAuszJ4JIowv:-OAXwpunAAuszJ4JIoww:bylakmy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f489804b5ef373bfc405b6049f1f3e1b8fa52bd0/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L103)):*
> ```Go
> 		id, found := zonepb.NamedZones[namedZone]
> 		if !found {
> 			return nil, fmt.Errorf("%q is not a built-in zone", string(zs.NamedZone))
> ```

If this is unexpected than lets return an assertion? 


<!-- Sent from Reviewable.io -->

annrpom commented:
reviewable is breaking on me; so i'll reply here:

> Does nil work with protoutil.Clone (that's the only caller of this function)? Or does it matter.

ah good point; i put the `zonepb.DefaultSystemZoneConfigRef()` value down, but the future we could make this more flexible if needed

> I'm not sure when the 25_1 variable will be added, so I'm fine with the TODO you have for now if you want to get this in sooner.

Ok I left a TODO here -- was gonna let it sit, but might as well merge

> If this is unexpected than lets return an assertion?

It's from a user doing something like `alter range faizan configure zone ...`; the DSC has https://github.com/cockroachdb/cockroach/blob/c68c559859be738efead9971f5e11f62a8c69d06/pkg/config/zonepb/zone.go#L152

annrpom commented:
TFTRs! ('-')7

bors r+
